src/startSupergraph.ts

```diff
@@ -0,0 +1,147 @@
+        if (config.configFile) {
+            productsSubgraph = await startProductsSubgraphUsingPm2(config.configFile);
+        }
+
+        const proc = execa("pm2", [
+            "start",
+            "supergraph.config.js",
+        ]);
+        proc.stdout.pipe(writeableDebugStream(pm2Debug));
+        proc.stderr.pipe(writeableDebugStream(pm2Debug));
+
+        if (productsSubgraph != null && productsSubgraph.exitCode !== 0) {
+            throw new Error("Products subgraph did not start successfully");
+        }
+
+        await proc;
+        if (proc.exitCode !== 0) {
+            throw new Error("Subgraphs did not start successfully");
+        }
+    
+        const started = await composeDevSupergraph(config.endpoint, config.schemaFile);
+        pm2Logs.cancel();
+    
+        if (started) {
+            return async () => {
+                pm2Debug(`\n***********************\nStopping supergraph...\n***********************\n\n`);
+                await shutdownSupergraphUsingPm2();
+            };
+        } else {
+            throw new Error("Supergraph did not start successfully");
+        }
+    } catch(err) {
+        await shutdownSupergraphUsingPm2();
+        throw err;
+    }
+}
+
+/**
+ * Flushes PM2 logs and then stops all PM2 processes (including daemon).
+ */
+async function shutdownSupergraphUsingPm2() {
+    await execa("pm2", ["flush"]);
+    await execa("pm2", ["kill"]);
+}
+
+async function startProductsSubgraphUsingPm2(configFile: string) {
+    // start products subgraph
+    pm2Debug(`Starting products subgraph...`);
+    const productsSubgraph = execa("pm2", [
+        "start",
+        configFile,
+    ]);
+
+    productsSubgraph.stdout.pipe(writeableDebugStream(pm2Debug));
+    productsSubgraph.stderr.pipe(writeableDebugStream(pm2Debug));
+    return productsSubgraph;
+}
+
+async function startSupergraphUsingDocker(config: DockerConfig) {
+    await composeSupergraph(config.schemaFile, config.path ?? "", config.port ?? "4001");
+    dockerDebug(`\n***********************\nStarting supergraph using Docker Compose...\n***********************\n\n`);
+
+    const proc = execa("docker", [
+        "compose",
+        "-f",
+        "docker-compose.yaml",
+        "-f",
+        config.composeFile,
+        "up",
+        "--build",
+        "--detach",
+    ]);
+
+    proc.stdout.pipe(writeableDebugStream(dockerDebug));
+    proc.stderr.pipe(writeableDebugStream(dockerDebug));
+
+    await proc;
+    if (proc.exitCode !== 0) {
+        throw new Error("docker-compose did not start successfully");
+    }
+
+    const started = await healthcheckRouter();
+    if (started) {
+        return async () => {
+            pm2Debug(`\n***********************\nStopping supergraph...\n***********************\n\n`);
+            await execa("docker-compose", ["down", "--remove-orphans", "-v"]);
+        };
+    } else {
+        throw new Error("Supergraph did not start successfully");
```

Comment:
Do we need to cleanup / `docker compose down` in this case?

Reply:
we never did in the original code and unsure if we need it but a good idea to be explicit